File: archive/WildstarStudio/Matrix.cpp
#include "StdAfx.h"
#include "Matrix.h"

Matrix::Matrix() {
	setIdentity();
}

void Matrix::setIdentity() {
	for (int i = 0; i < 4; ++i) {
		for (int j = 0; j < 4; ++j) {
			if (i == j)
				m[i][j] = 1;
			else
				m[i][j] = 0;
		}
	}
}

Matrix Matrix::operator * (const Matrix& rhs) const {
	Matrix ret;
	for (int i = 0; i < 4; ++i) {
		for (int j = 0; j < 4; ++j) {
			ret.m[i][j] =
				m[i][0] * rhs.m[0][j] +
				m[i][1] * rhs.m[1][j] +
				m[i][2] * rhs.m[2][j] +
				m[i][3] * rhs.m[3][j];
		}
	}
	return ret;
}

Vector3 Matrix::operator * (const Vector3& rhs) const {
	Vector3 ret;
	ret.X = rhs.X * m11 + rhs.Y * m12 + rhs.Z * m13 + m14;
	ret.Y = rhs.X * m21 + rhs.Y * m22 + rhs.Z * m23 + m24;
	ret.Z = rhs.X * m31 + rhs.Y * m32 + rhs.Z * m33 + m34;

	float w = 1.0f / (rhs.X * m41 + rhs.Y * m42 + rhs.Z * m43 + m44);

	ret *= w;

	return ret;
}

Matrix Matrix::transposed() const {
	Matrix ret = *this;
	ret.transpose();
	return ret;
}

void Matrix::transpose() {
	for (int i = 0; i < 4; ++i) {
		for (int j = 0; j < i; ++j) {
			float tmp = m[j][i];
			m[j][i] = m[i][j];
			m[i][j] = tmp;
		}
	}
}

float Matrix::determinant() {
	return m11 * m22 * m33 * m44 + m11 * m23 * m34 * m42 + m11 * m24 * m32 * m43
		+ m12 * m21 * m34 * m43 + m12 * m23 * m31 * m44 + m12 * m24 * m33 * m41
		+ m13 * m21 * m32 * m44 + m13 * m22 * m34 * m41 + m13 * m24 * m31 * m42
		+ m14 * m21 * m33 * m42 + m14 * m22 * m31 * m43 + m14 * m23 * m32 * m41
		- m11 * m22 * m34 * m43 - m11 * m23 * m32 * m44 - m11 * m24 * m33 * m42
		- m12 * m21 * m33 * m44 - m12 * m23 * m34 * m41 - m12 * m24 * m31 * m43
		- m13 * m21 * m34 * m42 - m13 * m22 * m31 * m44 - m13 * m24 * m32 * m41
		- m14 * m21 * m32 * m43 - m14 * m22 * m33 * m41 - m14 * m23 * m31 * m42;
}

bool Matrix::invert() {
	float det = determinant();
	if (det == 0.0f)
		return false;

	float invDet = 1.0f / det;

	__declspec(thread) static float b[4][4];

	b[0][0] = m22 * m33 * m44 + m23 * m34 * m42 + m24 * m32 * m43 - m22 * m34 * m43 - m23 * m32 * m44 - m24 * m33 * m42;
	b[0][1] = m12 * m34 * m43 + m13 * m32 * m44 + m14 * m33 * m42 - m12 * m33 * m44 - m13 * m34 * m42 - m14 * m32 * m43;
	b[0][2] = m12 * m23 * m44 + m13 * m24 * m42 + m14 * m22 * m43 - m12 * m24 * m43 - m13 * m22 * m44 - m14 * m23 * m42;
	b[0][3] = m12 * m24 * m33 + m13 * m22 * m34 + m14 * m23 * m32 - m12 * m23 * m34 - m13 * m24 * m32 - m14 * m22 * m33;

	b[1][0] = m21 * m34 * m43 + m23 * m31 * m44 + m24 * m33 * m41 - m21 * m33 * m44 - m23 * m34 * m41 - m24 * m31 * m43;
	b[1][1] = m11 * m33 * m44 + m13 * m34 * m41 + m14 * m31 * m43 - m11 * m34 * m43 - m13 * m31 * m44 - m14 * m33 * m41;
	b[1][2] = m11 * m24 * m43 + m13 * m21 * m44 + m14 * m23 * m41 - m11 * m23 * m44 - m13 * m24 * m41 - m14 * m21 * m43;
	b[1][3] = m11 * m23 * m34 + m13 * m24 * m31 + m14 * m21 * m33 - m11 * m24 * m33 - m13 * m21 * m34 - m14 * m23 * m31;

	b[2][0] = m21 * m32 * m44 + m22 * m34 * m41 + m24 * m31 * m42 - m21 * m34 * m42 - m22 * m31 * m44 - m24 * m32 * m41;
	b[2][1] = m11 * m34 * m42 + m12 * m31 * m44 + m14 * m32 * m41 - m11 * m32 * m44 - m12 * m34 * m41 - m14 * m32 * m42;
	b[2][2] = m11 * m22 * m44 + m12 * m24 * m41 + m14 * m21 * m42 - m11 * m24 * m42 - m12 * m21 * m44 - m14 * m22 * m41;
	b[2][3] = m11 * m24 * m32 + m12 * m21 * m34 + m14 * m22 * m31 - m11 * m22 * m34 - m12 * m24 * m31 - m14 * m21 * m32;

	b[3][0] = m21 * m33 * m42 + m22 * m31 * m43 + m23 * m32 * m41 - m21 * m32 * m43 - m22 * m33 * m41 - m23 * m31 * m42;
	b[3][1] = m11 * m32 * m43 + m12 * m33 * m41 + m13 * m31 * m42 - m11 * m33 * m42 - m12 * m31 * m43 - m13 * m32 * m41;
	b[3][2] = m11 * m23 * m42 + m12 * m21 * m43 + m13 * m22 * m41 - m11 * m22 * m43 - m12 * m23 * m41 - m13 * m21 * m42;
	b[3][3] = m11 * m22 * m33 + m12 * m23 * m31 + m13 * m21 * m32 - m11 * m23 * m32 - m12 * m21 * m33 - m13 * m22 * m31;

	for (int i = 0; i < 4; ++i) {
		for (int j = 0; j < 4; ++j)
			m[i][j] = b[i][j] * invDet;
	}

	return true;
}

Matrix Matrix::multiply(const Matrix& lhs, const Matrix& rhs) {
	Matrix ret;

	for (int i = 0; i < 4; ++i) {
		for (int j = 0; j < 4; ++j) {
			ret.m[i][j] =
				lhs.m[i][0] * rhs.m[0][j] +
				lhs.m[i][1] * rhs.m[1][j] +
				lhs.m[i][2] * rhs.m[2][j] +
				lhs.m[i][3] * rhs.m[3][j];
		}
	}

	return ret;
}

Matrix Matrix::scale(float x, float y, float z) {
	Matrix ret;
	ret.m11 = x;
	ret.m22 = y;
	ret.m33 = z;

	return ret;
}

Matrix Matrix::translation(float x, float y, float z) {
	Matrix ret;
	ret.m14 = x;
	ret.m24 = y;
	ret.m34 = z;

	return ret;
}

Matrix Matrix::translation(const Vector3& v) {
	return translation(v.X, v.Y, v.Z);
}

Matrix Matrix::rotation(float angle, const Vector3& axis) {
	angle = (angle * 3.141592f) / 180.0f;
	float cosv = cosf(angle);
	float sinv = sinf(angle);
	float ux = axis.X;
	float uy = axis.Y;
	float uz = axis.Z;
	float ux2 = ux * ux;
	float uy2 = uy * uy;
	float uz2 = uz * uz;

	Matrix ret;
	ret.m11 = cosv + ux2 * (1.0f - cosv);
	ret.m12 = ux * uy * (1 - cosv) - uz * sinv;
	ret.m13 = ux * uz * (1 - cosv) + uy * sinv;
	ret.m21 = uy * ux * (1 - cosv) + uz * sinv;
	ret.m22 = cosv + uy2 * (1 - cosv);
	ret.m23 = uy * uz * (1 - cosv) - ux * sinv;
	ret.m31 = uz * ux * (1 - cosv) - uy * sinv;
	ret.m32 = uz * uy * (1 - cosv) + ux * sinv;
	ret.m33 = cosv + uz2 * (1 - cosv);

	return ret;
}

Matrix Matrix::rotationAxis(const Vector3& v) {
	return rotation(v.X, Vector3::UnitX) * rotation(v.Y, Vector3::UnitY) * rotation(v.Z, Vector3::UnitZ);
}

Matrix Matrix::rotationCenter(float angle, const Vector3& axis, const Vector3& center) {
	Matrix toCenter = Matrix::translation(Vector3::negate(center));
	Matrix fromCenter = Matrix::translation(center);
	Matrix rotation = Matrix::rotation(angle, axis);

	Matrix ret;
	ret = Matrix::multiply(ret, fromCenter);
	ret = Matrix::multiply(ret, rotation);
	ret = Matrix::multiply(ret, toCenter);

	return ret;
}

Matrix Matrix::ortho(float left, float top, float right, float bottom, float zNear, float zFar) {
	Matrix ret;
	ret.m11 = 2.0f / (right - left);
	ret.m22 = 2.0f / (top - bottom);
	ret.m33 = -2.0f / (zFar - zNear);
	ret.m14 = -1.0f * ((right + left) / (right - left));
	ret.m24 = -1.0f * ((top + bottom) / (top - bottom));
	ret.m34 = -1.0f * ((zFar + zNear) / (zFar - zNear));

	return ret;
}

Matrix Matrix::perspective(float fovy, float aspect, float zNear, float zFar) {
	Matrix ret;
	float top = zNear * tanf(fovy * 3.141592f / 360.0f);
	float bottom = -top;
	float left = bottom * aspect;
	float right = top * aspect;
	ret.m11 = 2.0f * zNear / (right - left);
	ret.m13 = (right + left) / (right - left);
	ret.m22 = 2.0f * zNear / (top - bottom);
	ret.m23 = (top + bottom) / (top - bottom);
	ret.m33 = -(zFar + zNear) / (zFar - zNear);
	ret.m34 = -2.0f * zFar * zNear / (zFar - zNear);
	ret.m43 = -1;
	ret.m44 = 0.0f;

	return ret;
}

Matrix Matrix::lookAt(const Vector3& eye, const Vector3& at, const Vector3& up) {
	Vector3 forward = eye - at;
	forward.normalize();
	Vector3 side = Vector3::negate(forward.cross(up));
	side.normalize();

	Matrix ret;

	float floats[] = {
		side.X, up.X, forward.X, 0,
		side.Y, up.Y, forward.Y, 0,
		side.Z, up.Z, forward.Z, 0,
		-side.dot(eye), -up.dot(eye), -forward.dot(eye), 1

	};

	memcpy(ret._m, floats, sizeof(floats));

	ret.transpose();
	return ret;
}

File: archive/Online/js/Matrix.js
var Matrix = function () {
    this.values = new Float32Array(16);
    this.setIdentity();
}

Matrix.prototype.setIdentity = function () {
    for (var i = 0; i < 4; ++i) {
        for (var j = 0; j < 4; ++j) {
            this.values[i * 4 + j] = (i == j) ? 1 : 0;
        }
    }
}

Matrix.prototype.transpose = function () {
    for (var i = 0; i < 4; ++i) {
        for (var j = 0; j < 4; ++j) {
            var tmp = this.values[j * 4 + i];
            this.values[j * 4 + i] = this.values[i * 4 + j];
            this.values[i * 4 + j] = tmp;
        }
    }

    return this;
}

Matrix.scale = function (x, y, z) {
    var ret = new Matrix();
    ret.values[0] = x;
    ret.values[5] = y;
    ret.values[10] = z;

    return ret;
}

Matrix.translation = function (x, y, z) {
    var ret = new Matrix();
    ret.values[3] = x;
    ret.values[7] = y;
    ret.values[11] = z;

    return ret;
}

Matrix.rotation = function (angle, axis) {
    angle = (angle * Math.PI) / 180.0;
    var cosv = Math.cos(angle);
    var sinv = Math.sin(angle);
    var ux = axis.x;
    var uy = axis.y;
    var uz = axis.z;
    var ux2 = ux * ux;
    var uy2 = uy * uy;
    var uz2 = uz * uz;

    var ret = new Matrix();
    ret.values[0] = cosv + ux2 * (1.0 - cosv);
    ret.values[1] = ux * uy * (1 - cosv) - uz * sinv;
    ret.values[2] = ux * uz * (1 - cosv) + uy * sinv;
    ret.values[4] = uy * ux * (1 - cosv) + uz * sinv;
    ret.values[5] = cosv + uy2 * (1 - cosv);
    ret.values[6] = uy * uz * (1 - cosv) - ux * sinv;
    ret.values[8] = uz * ux * (1 - cosv) - uy * sinv;
    ret.values[9] = uz * uy * (1 - cosv) + ux * sinv;
    ret.values[10] = cosv + uz2 * (1 - cosv);
}

Matrix.rotationAxis = function (axis) {
    return Matrix.multiply(Matrix.multiply(Matrix.rotation(axis.x, Vector3.UnitX), Matrix.rotation(axis.y, Vector3.UnitY)), Matrix.rotation(axis.z, Vector3.UnitZ));
}

Matrix.perspective = function (fovy, aspect, zNear, zFar) {
    var ret = new Matrix();
    var top = zNear * Math.tan(fovy * Math.PI / 360.0);
    var bottom = -top;
    var left = bottom * aspect;
    var right = top * aspect;

    ret.values[0] = 2.0 * zNear / (right - left);
    ret.values[3] = (right + left) / (right - left);
    ret.values[6] = 2.0 * zNear / (top - bottom);
    ret.values[7] = (top + bottom) / (top - bottom);
    ret.values[10] = -(zFar + zNear) / (zFar - zNear);
    ret.values[11] = -2.0 * zFar * zNear / (zFar - zNear);
    ret.values[14] = -1;
    ret.values[15] = 0.0;

    return ret;
}

Matrix.ortho = function (left, top, right, bottom, zNear, zFar) {
    var ret = new Matrix();

    var lr = 1 / (left - right),
    bt = 1 / (bottom - top),
    nf = 1 / (zNear - zFar);
    ret.values[0] = -2 * lr;
    ret.values[1] = 0;
    ret.values[2] = 0;
    ret.values[3] = 0;
    ret.values[4] = 0;
    ret.values[5] = -2 * bt;
    ret.values[6] = 0;
    ret.values[7] = 0;
    ret.values[8] = 0;
    ret.values[9] = 0;
    ret.values[10] = 2 * nf;
    ret.values[11] = 0;
    ret.values[12] = (left + right) * lr;
    ret.values[13] = (top + bottom) * bt;
    ret.values[14] = (zFar + zNear) * nf;
    ret.values[15] = 1;

    return ret;
}

Matrix.lookAt = function (eye, at, up) {
    var forward = Vector3.subtract(eye, at);
    forward.normalize();
    var side = Vector3.negate(forward.cross(up));
    side.normalize();

    var ret = new Matrix();
    var floats = new Float32Array([
        side.x, up.x, forward.x, 0,
        side.y, up.y, forward.y, 0,
        side.z, up.z, forward.z, 0,
        -side.dot(eye), -up.dot(eye), -forward.dot(eye)
    ]);

    ret.values.set(floats);

    return ret;
}

Matrix.transposed = function (matrix) {
    var ret = new Matrix();
    for (var i = 0; i < 16; ++i) {
        ret.values[i] = matrix.values[i];
    }

    ret.transpose();
    return ret;
}

Matrix.multiply = function (matrix, vectorOrMatrix) {
    if (vectorOrMatrix instanceof Vector3) {
        var ret = new Vector3();
        var rhs = vectorOrMatrix;
        ret.x = rhs.x * matrix.values[0] + rhs.y * matrix.values[1] + rhs.z * matrix.values[2];
        ret.y = rhs.x * matrix.values[4] + rhs.y * matrix.values[5] + rhs.z * matrix.values[6];
        ret.z = rhs.x * matrix.values[8] + rhs.y * matrix.values[9] + rhs.z * matrix.values[10];

        var w = 1.0 / (rhs.x * matrix.values[12] + rhs.y * matrix.values[13] + rhs.z * matrix.values[14]);
        ret.x *= w;
        ret.y *= w;
        ret.z *= w;

        return ret;
    } else if (vectorOrMatrix instanceof Matrix) {
        var ret = new Matrix();
        var rhs = vectorOrMatrix;

        for (var i = 0; i < 4; ++i) {
            for (var j = 0; j < 4; ++j) {
                ret.values[i * 4 + j] =
                    matrix.values[i * 4] * rhs[j] +
                    matrix.values[i * 4 + 1] * rhs[4 + j] +
                    matrix.values[i * 4 + 2] * rhs[8 + j] +
                    matrix.values[i * 4 + 3] * rhs[12 + j];
            }
        }

        return ret;
    }
}